### DISCUSSION TITLE
Unexpected typing in __setitem__ with typevars

### DISCUSSION THREAD

**Discussion by Sharp-Eyes** [microsoft/pyright#6894] [Q&A] [answered]
It seems that `__setitem__` doesn't currently support using typevars to correlate key and value types.
A very basic example of what I mean would be something like
```py
class Foo:
    def __setitem__(self, key: type[T], value: T) -> None:
        ...

f = Foo()
f[int] = 123  # This is fine
f[int] = "abc"  # I'd expect this to error, but it doesn't
```
This does seem to work as intended with a intermediary generic class:
```py
class Key(typing.Generic[T]):
    def __init__(self, key: str):
        self.key = key

class Foo:
    def __setitem__(self, key: Key[T], value: T) -> None:
        ...

f = Foo()
k = Key[int]("abc")
f[k] = 123  # This is fine
f[k] = "abc"  # This causes a pyright error as expected
```
Is doing this without an intemediary class something that would be possible/plausible to support?

For the record, this was checked using:
- Pyright 1.1.344
- Pylance v2023.12.1

**Comment by erictraut** [comment, marked as answer]
I think you're misunderstanding how type variables work. When a type variable is used in a function or method, the type checker attempts to "solve" the type variable by finding some type that meets all of the constraints on that type variable.

Consider this simplified example:

```python
def func(a: T, b: T) -> T:
    ...

x = func(1, "")
reveal_type(x) # int | str
```

Note that the arguments passed to `a` and `b` are different types, but the constraint solver is able to find a type (`int | str`) that works with both arguments.

This is the same thing that happening in your first example above.

In your second example, the `Key` instance has already been specialized to `Key[int]` prior to calling `__setitem__`, and the type parameter for `Key` is invariant. This places a constraint on the solution, and the constraint solver is no longer able to find a type `T` that satisfies all the constraints.

**Reply by Sharp-Eyes** [reply]
Aha, that does make a lot of sense. Thank you for the explanation!
